### PR TITLE
feat: Expose additional data handlers as an argument in train

### PR DESCRIPTION
**Pull request by dushyantbehl**
### Description of the change

Add an argument in train to take in any user specified data handlers to be registered with the data preprocessor which can be invoked by a custom data config for advanced users.

### Was the PR tested

- [x] I have added >=1 unit test(s) for every new method I have added.
- [x] I have ensured all unit tests pass

### CODE REVIEW

**Comment by dushyantbehl** [comment]
@Abhishek-TAMU @willmj @ashokponkumar This PR is ready for review now and has unit tests for the additional data handler feature integrated.

File: tuning/data/data_processors.py [hunk 1]
```diff
@@ -46,9 +46,20 @@ def __init__(
         # Initialize other objects
         self.registered_handlers = {}
 
-    def register_data_handler(self, name: str, func: callable):
+    def register_data_handler(self, name: str, func: Callable):
+        if not isinstance(name, str) or not callable(func):
+            raise ValueError("Handlers should be of type Dict, str to callable")
+        logging.info("Registering handler %s passed by the user", name)
         self.registered_handlers[name] = func
```
Comment (ashokponkumar) [inline tuning/data/data_processors.py:53]: Do you want to return a warn message if a data handler of the same name exists before?

File: tuning/data/setup_dataprocessor.py [hunk 2]
```diff
@@ -179,14 +184,15 @@ def _process_raw_data_args(
...
+    additional_data_handlers: Dict[str, Callable] = None,
 ):
 
     # Create a data processor with default processor config
     default_processor_config = DataPreProcessorConfig()
     data_processor = get_datapreprocessor(
         processor_config=default_processor_config, tokenizer=tokenizer
     )
-
+    data_processor.register_data_handlers(additional_data_handlers)
```
Comment (ashokponkumar) [inline tuning/data/setup_dataprocessor.py:195]: Why not pass the additional handlers also to the get_datapreprocessor?

File: tuning/data/data_processors.py [hunk 3]
```diff
@@ -238,19 +249,12 @@ def process_dataset_configs(
...
-
 def get_datapreprocessor(
     processor_config: DataPreProcessorConfig, tokenizer: AutoTokenizer
 ) -> DataPreProcessor:
     processor = DataPreProcessor(
         processor_config=processor_config,
         tokenizer=tokenizer,
     )
-    autoregister_available_handlers(processor)
+    processor.register_data_handlers(AVAILABLE_DATA_HANDLERS)
```
Comment (ashokponkumar) [inline tuning/data/data_processors.py:259]: If this function is called twice, won't this be added twice?

**Review by ashokponkumar — APPROVED** [review]
Few nits

**Review by ashokponkumar — CHANGES_REQUESTED** [review]
Requested few nits

File: tuning/data/setup_dataprocessor.py (hunk 2 above)
Comment (dushyantbehl) [reply tuning/data/setup_dataprocessor.py:195]: done

File: tuning/data/data_processors.py (hunk 3 above)
Comment (dushyantbehl) [reply tuning/data/data_processors.py:259]: yes..added a warning if the code updates an existing handler

Comment (dushyantbehl) [reply tuning/data/data_processors.py]: yes..added a warning as you mentioned for this scenario

File: tuning/data/setup_dataprocessor.py (hunk 2 above)
Comment (dushyantbehl) [reply tuning/data/setup_dataprocessor.py:196]: done

File: tuning/data/data_processors.py (hunk 1 above)
Comment (dushyantbehl) [reply tuning/data/data_processors.py:58]: added.

File: tuning/data/data_processors.py (hunk 3 above)
Comment (ashokponkumar) [reply tuning/data/data_processors.py]: I think for default handlers we should just add it upfront or have a flag to ensure it does not get added twice irrespective of how many times it is called.

**Review by ashokponkumar — DISMISSED** [review]

File: tests/test_sft_trainer.py [hunk 4]
```diff
@@ -1124,3 +1134,88 @@ def test_pretokenized_dataset_wrong_format():
...
+@pytest.mark.parametrize(
+    "additional_handlers",
+    [
+        "thisisnotokay",
+        [],
+        {lambda x: {"x": x}: "notokayeither"},
+        {"thisisfine": "thisisnot"},
+    ],
+)
+def test_run_with_bad_additional_data_handlers(additional_handlers):
```
Comment (willmj) [inline tests/test_sft_trainer.py:1148]: nit: docstrings for all added test cases

**Review by willmj — APPROVED** [review]
LGTM, but before merging let's add some docstrings to test cases to explain what they are testing

File: tests/test_sft_trainer.py (hunk 4 above)
Comment (dushyantbehl) [reply tests/test_sft_trainer.py:1155]: Added docstrings. Thanks @willmj

File: tests/test_sft_trainer.py [hunk 5]
```diff
@@ -1124,3 +1134,98 @@ def test_pretokenized_dataset_wrong_format():
...
+        train_args = copy.deepcopy(TRAIN_ARGS)
+        train_args.output_dir = tempdir
+
+        sft_trainer.train(
+            MODEL_ARGS,
+            DATA_ARGS,
+            train_args,
+            PEFT_PT_ARGS,
+            additional_data_handlers=None,
+        )
```
Comment (Abhishek-TAMU) [inline tests/test_sft_trainer.py:1184]: We can just add this line after this:

`_validate_training(tempdir)`

File: tests/test_sft_trainer.py [hunk 6]
```diff
@@ -1124,3 +1134,98 @@ def test_pretokenized_dataset_wrong_format():
...
+        data_args.data_config_path = data_config_path
+        data_args.dataset_text_field = "custom_formatted_field"
+
+        sft_trainer.train(
+            MODEL_ARGS,
+            DATA_ARGS,
+            train_args,
+            PEFT_PT_ARGS,
+            additional_data_handlers={TEST_HANDLER: test_handler},
+        )
```
Comment (Abhishek-TAMU) [inline tests/test_sft_trainer.py:1231]: Here also, we can just add this line after this:

`_validate_training(tempdir)`

**Review by Abhishek-TAMU — COMMENTED** [review]
Looks good to me, other than just few changes!

**Comment by Abhishek-TAMU** [comment]
@willmj @dushyantbehl I just pushed the PR changes suggested by myself and some doc string changes suggested by Will. Looks good to merge now.

**Review by ashokponkumar — APPROVED** [review]

**Pull request merged**